jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/CharsetStringBuilderTest.java

```diff
@@ -71,34 +75,109 @@ public static Stream<Charset> charsets()
+
+        // using only append(byte) recreate the input
+        CharsetStringBuilder builder = CharsetStringBuilder.forCharset(charset);
+        for (byte b : buf)
+        {
+            builder.append(b);
+        }
 
-        builder.append(charset.encode("34"));
+        assertThat(builder.build(), is(input));
+    }
 
-        encoded = charset.encode("abc");
-        int offset = encoded.remaining();
-        encoded = charset.encode("abc56");
-        int length = encoded.remaining() - offset;
-        encoded = charset.encode("abc56xyz");
-        byte[] bytes = new byte[1028];
-        encoded.get(bytes, 0, encoded.remaining());
-        builder.append(bytes, offset, length);
+    @ParameterizedTest
+    @MethodSource("charsets")
+    public void testAppendByteOffsetLengthOnly(Charset charset) throws Exception
+    {
+        String input = "123456789ABC";
 
-        encoded = charset.encode("abc78xyz");
-        encoded.position(offset);
-        encoded.limit(offset + length);
-        builder.append(encoded);
+        // Generate a byte buffer encoded with the provided charset of the input String.
+        byte[] buf = input.getBytes(charset);
 
-        builder.append("9A", 0, 2);
-        builder.append("xyzBCpqy", 3, 2);
+        // using only append(byte, offset, length) recreate the input
+        CharsetStringBuilder builder = CharsetStringBuilder.forCharset(charset);
+        int sliceSize = 3;
+        int offset = 0;
+        while (offset < buf.length)
+        {
+            int len = Math.min(sliceSize, buf.length - offset);
+            builder.append(buf, offset, len);
+            offset += sliceSize;
+        }
+
+        assertThat(builder.build(), is(input));
+    }
 
-        assertThat(builder.build(), is("123456789ABC"));
+    @ParameterizedTest
+    @MethodSource("charsets")
+    public void testAppendCharOnly(Charset charset) throws Exception
+    {
+        String input = "123456789ABC";
+
+        // using only append(char) recreate the input
+        CharsetStringBuilder builder = CharsetStringBuilder.forCharset(charset);
+        for (char c : input.toCharArray())
+        {
+            builder.append(c);
+        }
+
+        assertThat(builder.build(), is(input));
+    }
+
+    @ParameterizedTest
+    @MethodSource("charsets")
+    public void testAppendCharSequenceOffsetLengthOnly(Charset charset) throws Exception
+    {
+        String input = "123456789ABC";
+
+        // using only append(CharSequence, offset, length) recreate the input
+        CharsetStringBuilder builder = CharsetStringBuilder.forCharset(charset);
+        char[] chars = input.toCharArray();
+        int sliceSize = 3;
+        int offset = 0;
+        while (offset < chars.length)
+        {
+            int len = Math.min(sliceSize, chars.length - offset);
+            builder.append(input, offset, len);
+            offset += sliceSize;
+        }
+
+        assertThat(builder.build(), is(input));
```

Comment:
How about adding this test:

```suggestion
        assertThat(builder.build(), is(input));
    }

    public static Stream<Arguments> japaneseCharsetsTests()
    {
        List<Arguments> args = new ArrayList<>();

        for (Charset charset : List.of(
            StandardCharsets.UTF_8,
            StandardCharsets.UTF_16,
            Charset.forName("Shift_JIS"),
            Charset.forName("EUC-JP")
        ))
        {
            for (String string : List.of("ホ", "カタカナ", "ｶﾞｯﾂﾎﾟｰｽﾞ"))
                args.add(Arguments.of(charset, string));
        }
        return args.stream();
    }

    @ParameterizedTest
    @MethodSource("japaneseCharsetsTests")
    public void testJapaneseCharsetsMixedEncodings(Charset charset, String string) throws Exception
    {
        CharsetStringBuilder builder = CharsetStringBuilder.forCharset(charset);

        byte[] bytes = string.getBytes(charset);
        for (byte b : bytes)
        {
            if (b > 'a' && b < 'z' || b > 'A' && b < 'Z' || b > '0' && b < '9')
                builder.append((char)b);
            else
                builder.append(b);
        }
        assertThat(builder.build(), is(string));
```